### PR TITLE
Fix RichAudience Bidder Bug

### PR DESCRIPTION
**Pull request by AntoxaAntoxic**

### CODE REVIEW

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 1]
```diff
@@ -91,23 +96,25 @@ private static void validateRequest(BidRequest bidRequest) throws PreBidExceptio
...
-            return new URL(ObjectUtil.getIfNotNull(bidRequest.getSite(), Site::getPage));
-        } catch (MalformedURLException e) {
-            return null;
-        }
-    }
-
+    //todo: what if imp is not a banner and not a video?
     private static void validateImp(Imp imp) throws PreBidException {
-        if (!isBannerSizesPresent(imp.getBanner())) {
+        //todo: does it make sense to collect all the errors like in Go?
```
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:101]: this comment is redundant, to be removed

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 2]
```diff
@@ -177,24 +209,34 @@ private List<BidderBid> extractBids(BidResponse bidResponse, BidRequest bidReque
...
-                return BidType.banner;
-            }
+    private static BidderBid makeBid(BidResponse bidResponse, Map<String, Imp> impMap, Bid bid) {
+        final Imp imp = impMap.get(bid.getImpid());
+        final BidType bidType = resolveBidType(imp);
+        final Bid.BidBuilder builder = bid.toBuilder();
+
+        if (bidType == BidType.video) {
+            builder.w(imp.getVideo().getW());
+            builder.h(imp.getVideo().getH());
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:232]: Let's make some defensive programming things)) and make same checks if imp.getVideo() is null

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 3]
```diff
@@ -91,23 +96,25 @@ private static void validateRequest(BidRequest bidRequest) throws PreBidExceptio
...
 
-    private static URL extractUrl(BidRequest bidRequest) {
-        try {
-            return new URL(ObjectUtil.getIfNotNull(bidRequest.getSite(), Site::getPage));
-        } catch (MalformedURLException e) {
-            return null;
-        }
-    }
-
+    //todo: what if imp is not a banner and not a video?
```
[the remark this answers is not in the document]
Comment (AntoxaAntoxic) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:99]: I mean it's a config, one of the feature of the config is being changed

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java (hunk 2 above)
Comment (SerhiiNahornyi) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:241]: agree

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (SerhiiNahornyi) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java]: This config almost always comes to us with bidder change, it is not the part that is changed in prod

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 4]
```diff
@@ -177,24 +218,34 @@ private List<BidderBid> extractBids(BidResponse bidResponse, BidRequest bidReque
...
+
         return bidResponse.getSeatbid().stream()
                 .filter(Objects::nonNull)
-                .map(SeatBid::getBid)
+                .flatMap(seatBid -> seatBid.getBid().stream())
                 .filter(Objects::nonNull)
-                .flatMap(Collection::stream)
-                .map(bid -> BidderBid.of(bid,
-                        resolvedBidType(bid.getImpid(), bidRequest.getImp()), bidResponse.getCur()))
+                //todo: is it fine to return a bid with unknown bid type? (as the PBS Go does)
```
Comment (AntoxaAntoxic) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:228]: @SerhiiNahornyi any comments on that?

Comment (And1sS) [reply src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:228]: Yes, it's fine, you can check `ResponseBidValidator.java` to double check this.
But, PBS Go actually returns bid type for correct imps.

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 5]
```diff
@@ -118,12 +122,63 @@ private ExtImpRichaudience parseImpExt(Imp imp) throws PreBidException {
...
+        final boolean isSecure = urlOptional.map(URL::getProtocol).map(HTTPS::equals).orElse(false);
+        final Imp modifiedImp = modifyImp(imp, extImp, isSecure);
+        final Site modifiedSite = urlOptional
+                .map(url -> modifySite(originalSite, imp.getTagid(), url))
+                .orElseGet(() -> modifySite(originalSite, imp.getTagid()));
+        final App modifiedApp = modifyApp(request.getApp(), imp.getTagid());
+        final boolean isTest = BooleanUtils.isTrue(extImp.getTest());
+
+        return request.toBuilder()
+                .imp(List.of(modifiedImp))
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:138]: `Collections.singletonList()`

File: src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java [hunk 6]
```diff
@@ -57,31 +64,28 @@ public RichaudienceBidder(String endpointUrl, JacksonMapper mapper) {
...
             return Result.withError(BidderError.badInput(e.getMessage()));
         }
 
-        return Result.withValues(httpRequests);
+        for (Imp imp : request.getImp()) {
+            try {
+                validateImp(imp);
+                final ExtImpRichaudience extImp = parseImpExt(imp);
+                final BidRequest modifiedBidRequest = makeRequest(request, imp, extImp);
+                httpRequests.add(makeHttpRequest(modifiedBidRequest, Set.of(imp.getId())));
```
Comment (And1sS) [inline src/main/java/org/prebid/server/bidder/richaudience/RichaudienceBidder.java:81]: `Collections.singleton()`

**Review by And1sS — CHANGES_REQUESTED** [review]

**Review by SerhiiNahornyi — APPROVED** [review]

**Pull request merged**